src/server/pfs/server/driver.go

```diff
@@ -2186,25 +2265,77 @@ func (d *driver) getFile(ctx context.Context, file *pfs.File, offset int64, size
 	if err != nil {
 		return nil, err
 	}
-
 	paths, err := tree.Glob(file.Path)
 	if err != nil {
 		return nil, err
 	}
+	var objects []*pfs.Object
+	var totalSize int64
+	var sortedPaths []string
+	foundDirectoryNode := false
 	for path, node := range paths {
-		if node.FileNode == nil {
-			delete(paths, path)
+		sortedPaths = append(sortedPaths, path)
+		if node.DirNode != nil {
+			foundDirectoryNode = true
 		}
 	}
+	if !foundDirectoryNode {
+		// e.g. this was a request for a single file e.g. /foo/bar.txt
+		dirNodePaths, err := tree.Glob(filepath.Dir(file.Path))
+		if err != nil {
+			return nil, err
+		}
+		if len(dirNodePaths) != 1 {
+			return nil, fmt.Errorf("invalid set of directory node paths - found %v expected 1 - this is probably a bug", len(dirNodePaths))
+		}
+		for key, node := range dirNodePaths {
+			paths[key] = node
+			sortedPaths = append(sortedPaths, key)
+		}
+	}
```

Comment:
I don't total understand what this block is doing. Why do we error if `len(dirNodePaths) != 1`? That doesn't seem like correct semantics, ie if someone did `get-file */0000000`, which gets the first row of each table at the top level, it seems like this would error if there were multiple directories, but that's a perfectly valid `get-file`.

At a high level I'm a bit confused about why you think you need separate code paths for the case where your glob pattern matches a directory and the case where it doesn't. It seems like you should be able to handle everything with the stack approach.